### PR TITLE
[Caffe2]Enable fusion for IDEEP in optimizeForIdeep

### PR DESCRIPTION
**Pull request by gujinghui**
Enable fusion for IDEEP in optimizeForIdeep, including Conv+ReLU, Conv+Sum, Conv+Sum+ReLU, Conv+BN, and pre-convert filter format

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai 
pls review

File: caffe2/ideep/operators/conv_fusion_op.cc [hunk 1]
```diff
@@ -75,25 +73,10 @@ class IDEEPConvFusionOp final : public IDEEPConvPoolOpBase {
         "*",
         group_);
 
-    bool weights_changed =
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_fusion_op.cc]: Why do we remove this?

File: caffe2/ideep/operators/conv_op.cc [hunk 2]
```diff
@@ -36,21 +34,6 @@ class IDEEPConvOp final : public IDEEPConvPoolOpBase {
         "*",
         group_);
 
-    bool weights_changed =
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_op.cc]: Same here

File: caffe2/opt/optimize_ideep.cc [hunk 3]
```diff
@@ -2,42 +2,404 @@
...
+  op->set_type("ConvFusion");
+  auto* arg = op->add_arg();
+  arg->set_name("fusion_type");
+  arg->set_i(fusion_type);
+}
 
-    // We only want to fuse for IDEEP convs
-    if (op->device_option().device_type() != DeviceType::IDEEP) {
-      return false;
+bool fuseConvBNHelperForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: `ConvBN` fusion isn't IDEEP specific. And I believe it's done in 
https://github.com/pytorch/pytorch/blob/master/caffe2/opt/fusion.cc#L17

File: caffe2/opt/optimize_ideep.cc [hunk 4]
```diff
@@ -2,42 +2,404 @@
...
+
+  return false;
+}
+
+void fuseConvBNForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
+  while (fuseConvBNHelperForIdeep(nn, ws)) {
+  }
+}
+
+void fuseConvSumForIdeep(repr::NNModule* nn, caffe2::Workspace* ws) {
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: I still don't understand what's the use case for conv+sum fusion in inference. Could you explain?

File: caffe2/opt/optimize_ideep.cc [hunk 5]
```diff
@@ -2,42 +2,404 @@
...
+    auto newOutputTensor = util::make_unique<repr::Tensor>(inputName);
+    auto newOutput = nn->dataFlow.createNode(
+        unique_dyn_cast<repr::NeuralNetData>(newOutputTensor));
+    nn->dataFlow.replaceNode(convOutput, newOutput);
+
+    nn->dataFlow.deleteNode(convOutput);
+  }
+}
+
+void preConvertFiltersFormat(repr::NNModule* nn, caffe2::Workspace* ws) {
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: What does this do?

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/ideep/operators/conv_fusion_op.cc]: The filter has been converted to optimized format already in optimizeForIdeep.
Nothing needs to do here for inference.

File: caffe2/ideep/operators/conv_op.cc (hunk 2 above)
Comment (gujinghui) [reply caffe2/ideep/operators/conv_op.cc]: The filter has been converted to optimized format already in optimizeForIdeep.
Nothing needs to do here for inference.

File: caffe2/opt/optimize_ideep.cc (hunk 3 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: Yes. But difference tensor type and tensor API. It's difficult to invoke existing ConvBN fusion directly for Ideep tensors.

File: caffe2/opt/optimize_ideep.cc (hunk 5 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: Convert filter formats to optimized one just one time before runing.

File: caffe2/opt/optimize_ideep.cc (hunk 4 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: Used to fuse residual block, for example in resnet50.
I'm afraid I did not understand your question well. Could you descrbe more on your concern?
untitled

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (yinghai) [reply caffe2/ideep/operators/conv_fusion_op.cc]: This means we have to run optimizeForIdeep in our flow. I prefer that we keep this, because if you have run optimization, this is a no-op but if we somehow missed it, it will implicitly incur large overhead.

File: caffe2/python/ideep/convfusion_op_test.py [hunk 6]
```diff
@@ -189,6 +190,34 @@ def test_convolution_sum_fusion(self, stride, pad, kernel, size,
             print(S0.flatten())
             print(np.max(np.abs(S1 - S0)))
             self.assertTrue(False)
+
+        # Auto fusion for Conv + Sum
+        workspace.ResetWorkspace()
```
Comment (yinghai) [inline caffe2/python/ideep/convfusion_op_test.py]: I still don't understand how this can be used in inference and for Conv+Sum fusion, output will rewrite input 'S0'. If you run this op twice, you will get 2 different results as you input has changed. Please check https://github.com/intel/ideep/issues/8 for more discussion.

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (jgong5) [reply caffe2/ideep/operators/conv_fusion_op.cc]: @yinghai Since the weight format optimization can be applied to both training and inference, another option is that we optimize the format in the default path other than explicitly in optimizeForIdeep() function. Then we do not need to check the "weights_changed" explicitly in convolution op without incurring the "large overhead". What do you think?

Comment (jgong5) [reply caffe2/ideep/operators/conv_fusion_op.cc]: I meant the function preConvertFiltersFormat() used for weight format optimization.

File: caffe2/python/ideep/convfusion_op_test.py (hunk 6 above)
Comment (gujinghui) [reply caffe2/python/ideep/convfusion_op_test.py]: 1. training_mode flag changed the conv operator API for user, especially in training mode. Instead, optimizeForIdeep gives choice to user. :)

2. Yes. You're right. The input will be rewritten. But if the input is produced by other operator, instead of directly loaded by user, it should be OK in this case. Will add the check for it.

**Comment by gujinghui** [comment]
hi @yinghai 
Did you get our mail on this PR?

**Comment by gujinghui** [comment]
@yinghai  pls help review

@jgong5

**Comment by yinghai** [comment]
@gujinghui Sorry for the late reply. I'm working on verifying the fix for group convolution and next item is to check the memory increase issue. I'll get back to this one once the above outstanding issues are resolved as this one is not a blocker to anything. Makes sense?

**Comment by jgong5** [comment]
@yinghai Please let us know anything we can help to troubleshoot the memory footprint issue. I guess we can reproduce the problem with any model like ResNet-50, right?

**Comment by gujinghui** [comment]
@yinghai  As @jgong5 said, perhaps we can work together to figure out what happen on memory consumption issue?

**Comment by yinghai** [comment]
@jgong5 @gujinghui yeah, to start, just try resnet50 with ideep and mkl and check the difference.

**Comment by yinghai** [comment]
@gujinghui Actually do you have mask RCNN case to run? That might repo the memory issue better.

**Comment by gujinghui** [comment]
@yinghai 
rebased. pls review

File: cmake/Modules/FindMKL.cmake [hunk 7]
```diff
@@ -304,7 +304,7 @@ if (USE_MKL AND USE_IDEEP)
   
   if (MKLDNN_INCLUDE_DIR)
     list(APPEND IDEEP_INCLUDE_DIR ${MKLDNN_INCLUDE_DIR})
-    list(APPEND __ideep_looked_for ${MKLDNN_INCLUDE_DIR})
+    list(APPEND __ideep_looked_for MKLDNN_INCLUDE_DIR)
```
Comment (yinghai) [inline cmake/Modules/FindMKL.cmake]: Actually, why do we do this? What's the error msg you saw if you treated it as a variable?

Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: cmake log:

-- Could NOT find IDEEP (missing:  /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include)
CMake Error at cmake/Modules/FindMKL.cmake:360 (message):
  Did not find IDEEP files!

Comment (yinghai) [reply cmake/Modules/FindMKL.cmake]: This doesn't look right. You should be able to find `MKLDNN` headers there, right? Could you print out `__ideep_looked` to see what it's looking for?

Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: $ cmake -version
cmake version 3.5.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).

-------------------------------------------

**without this fix:
before __ideep_looked_for: /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include
-- Could NOT find IDEEP (missing:  /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include)
after __ideep_looked_for: /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include
CMake Error at cmake/Modules/FindMKL.cmake:362 (message):
  Did not find IDEEP files!

------------------------------------------

**with this fix:
before __ideep_looked_for: MKLDNN_INCLUDE_DIR
-- Found IDEEP: /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include
after __ideep_looked_for: MKLDNN_INCLUDE_DIR
-- Found IDEEP (include: /home/jgu21/workspace/src/intel_pytorch/third_party/ideep/include;/home/jgu21/workspace/src/intel_pytorch/third_party/ideep/mkl-dnn/include, lib: /home/jgu21/workspace/src/intel_pytorch/zz/lib/libmkldnn.so)

Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: the find_package_handle_standard_args shoulld need var list, other than string lists...

Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: BTW, with presnet FindMKL.cmake, the libmkldnn will be linked with gomp, not iomp5.
This should be not expected, and the performance is not garuanteed.

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (yinghai) [reply caffe2/ideep/operators/conv_fusion_op.cc]: Please don't remove this. I understand with the transformation, this step can be saved but it's not guaranteed that user will run transformation before running the nets. Opt is really an optional step.

File: caffe2/ideep/operators/conv_op.cc [hunk 8]
```diff
@@ -21,41 +19,23 @@ class IDEEPConvOp final : public IDEEPConvPoolOpBase {
     const auto& X = Input(INPUT);
     const auto& filter = Input(FILTER);
     auto* Y = Output(OUTPUT);
-    auto Y_dims = CalcOutputDims(X, filter.get_dim(0));
+    auto grouped = filter.is_grouped() ? 1 : 0;
+    auto Y_dims = CalcOutputDims(
```
Comment (yinghai) [inline caffe2/ideep/operators/conv_op.cc]: These changes don't seem to be related to this PR.

File: caffe2/ideep/operators/fully_connected_op.cc [hunk 9]
```diff
@@ -18,16 +20,39 @@ class IDEEPFullyConnectedOp final : public IDEEPOperator {
     const auto& filter = Input(FILTER);
     auto* Y = Output(OUTPUT);
 
+    auto newDims = [&](itensor::dims adims, size_t axis) {
```
Comment (yinghai) [inline caffe2/ideep/operators/fully_connected_op.cc]: And what's this change about?

File: cmake/Modules/FindMKL.cmake (hunk 7 above)
Comment (yinghai) [reply cmake/Modules/FindMKL.cmake]: > BTW, with presnet FindMKL.cmake, the libmkldnn will be linked with gomp, not iomp5.
This should be not expected, and the performance is not garuanteed.

This is intentional. User should be able to choose to use gomp or iomp, but absolutely not both of them. I have benched the performance of using gomp and iomp and they are similar to me. Note that C2 can be included as a library and used by other applications. Situation is that if they use gomp and unknowingly linked in iomp from C2, they can have a nasty coredump and it's very difficult for them to figure it out without context.

**Review by yinghai — CHANGES_REQUESTED** [review]
Thanks for pushing this. I have some comments that needs to be addressed. In addition, I think this PR is losing focus, as it is supposed to be transformation. However, I saw 
- Transformation
- Fixes in group size and FC
- Fixes in CMakefile

I suggest we split this PR into 3. Actually, let me take the CMakefile fix. And you can work on to split this into 2. Sounds good?

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (jgong5) [reply caffe2/ideep/operators/conv_fusion_op.cc]: @yinghai The concern is that this weight cache required the change of the API that creates the convolution operator, i.e. the "training_mode" flag which is off by default. Would this introduce the compatibility issue and require additional changes of the models for training? These models have to explicitly flag "training_mode" on, right?

**Comment by gujinghui** [comment]
This PR is being split to 2.
Pls refer to #9255 on op fusion part.

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (yinghai) [reply caffe2/ideep/operators/conv_fusion_op.cc]: There is no compatibility issue right now as IDEEP training path is brand-new. On the other hand, relying on optimization pass like this before doing inference introduces backward compatibility issue as previously we don't need this.

File: caffe2/ideep/operators/conv_op.cc (hunk 8 above)
Comment (yinghai) [reply caffe2/ideep/operators/conv_op.cc]: Split this out?

File: caffe2/ideep/operators/conv_fusion_op.cc (hunk 1 above)
Comment (jgong5) [reply caffe2/ideep/operators/conv_fusion_op.cc]: @yinghai I don't insist we should remove this in this PR. We can keep this weight cache for now. But I do think the API for creating models should be the same regardless of the backend devices the models run with, not a problem of whether a backend is new or not. With this "training_mode" flag, users have to explicitly modify their training models so as to run on MKL-DNN optimized backend, right?

**Comment by gujinghui** [comment]
@yinghai  This PR is not needed now. Will close it.
